src/clients.rs

```diff
@@ -257,20 +273,28 @@ impl StateClient for State {
             last_height = std::cmp::max(last_height, rsp.height);
 
             let note = Note::from_slice(&rsp.note)?;
-            let ownership = vk.owns(&note);
-            let hash = ownership.then(|| note.hash());
-            let note = ownership.then_some(note);
 
-            state.cache.insert(psk, rsp.height, (note, hash))?;
+            for (ssk, vk, psk) in addresses.iter() {
+                if let Some(insert) = vk.owns(&note).then(|| {
+                    state.cache.insert(
+                        psk,
+                        rsp.height,
+                        (note, note.gen_nullifier(ssk)),
+                    )
+                }) {
+                    insert?;
+                    break;
```

Comment:
Gotcha, but it's a bit odd imo

Why not simply `if vk.owns(&note) => { insert? and break }`